### PR TITLE
validation: allow only one transaction over 100kb.

### PR DESCRIPTION
**Pull request by chjj**
This adds a new post-HF consensus rule: only one transaction in a block is allowed to be above 100kb in _base size_. See #32.

### CODE REVIEW

**Comment by Buck-Rogers** [comment]
This has the potential to cause problems post HF as other clients do not have this rule.  I do not see a reasonable justification for it other then making more ways for the client to fail. Not good IMHO.

**Comment by opetruzel** [comment]
@Buck-Rogers What other clients exist that are already fully compatible with the 2M/8M SegWit consensus layer found in the SegWit2x HF? If all other clients must already upgrade for full compatibility, then adding this rule when they do so wouldn't be a problem, would it?  And, while the softfork aims for compatibility, the same is obviously not true of any hardfork.  If the HF is successful, SegWit2x becomes the reference client that all others would have to follow, in terms of consensus rulesets, so your response and concern confuse me a bit. Do you have a particular client in mind that is otherwise already compatible?

**Comment by earonesty** [comment]
@opetruzel spv clients don't have to know anything about block size limits.   so they won't need to update their code for the hard fork.

**Comment by opetruzel** [comment]
@earonesty 
True, but those wouldn't be creating or validating the blocks either, so how would this rule impact/affect them?

**Comment by sneurlax** [comment]
NACK, this is a change which should only be implemented after going through the BIP process ... or the BUIP process if they should wish to merge it.

**Comment by opetruzel** [comment]
@sneurlax 
Why would the formal BIP and BUIP processes be immediately relevant to this independent project?

**Comment by murchandamus** [comment]
@opetruzel: Consensus rules affect all Bitcoin participants, and are therefore not a matter of a single client.

**Comment by opetruzel** [comment]
@Xekyo 
I know that very well, and I personally wish the timeline wasn't so condensed that it prevents a more formal and lengthy review process. However, reality is what it is, and the goal here includes a hardfork with new rules. So, updating to accommodate those new rules, or not, is an inevitable decision that every other client dev team is going to have to make in the near future **if** SegWit2x is successful.

**Comment by jameshilliard** [comment]
> spv clients don't have to know anything about block size limits. so they won't need to update their code for the hard fork.

This means spv wallets are vulnerable to wipeout.

**Comment by Buck-Rogers** [comment]
@opetruzel 
> What other clients exist that are already fully compatible with the 2M/8M SegWit consensus layer found in the SegWit2x HF? 

About 41%+ of the miners running BU/XT/Classic/EC are compatible with segwit2x. These clients would ignore segwit txs and handle the HF just fine so they would follow any segwit2x chain. They would work both as a node and as a miner, they do not need to update to support segwit2x. So this change would affect them negatively as it may cause them to not follow this chain. Thus it is a NACK from me.

**Comment by opetruzel** [comment]
@Buck-Rogers 
For the sake of argument, let's assume that those miners are actually running BU, and not just signaling BU for political reasons. First, a very large percentage of that 41% belongs to companies that signed the NYA, so aren't they expected to switch to SegWit2x in July as part of their commitment? Second, why should the SegWit2x hardfork be designed to cater to their old consensus rules, and not the other way around? AFAIC, we should not concern ourselves with other clients that may need to update themselves to remain compatible with the SegWit2x hardfork consensus layer.  BU is no different than Core in that regard.

**Comment by Buck-Rogers** [comment]
@opetruzel I fully support the segwit2x agreement. However I will never mine or run a node that supports segwit transactions as I have a reasonable amount of contempt for segwit as a solution. These things are not mutually exclusive. I think you need to have a look at why core failed to get a compromise here, your attitude in your response & your lack of understanding of other positions mirrors the reasons why core failed to get consensus. Do you fully support segwit2x?

**Comment by opetruzel** [comment]
@Buck-Rogers 
I fully support the charter for SegWit2x, and my interpretation of said charter is a protocol that implements SegWit and 2M/8M blocks as the new consensus layer. I personally couldn't care less about maintaining ANY compatibility with your protest client that doesn't support SegWit 2M/8M, so perhaps I'm the wrong person to ask. Your personal crusade to continuously throw a wrench into the consensus mix for years to come is no concern of mine.

**Comment by earonesty** [comment]
@opetruzel a lightweight mobile client can, today, generate a 200k transaction with lots of small inputs.   it will not have to update for the hard fork, as long as the nodes it uses are updated... it will work fine.   likewise a hardware wallet only needs to know how to sign a tx, and doesn't need to know anything about block size limits.

changing transaction generation requires all of these clients to change.   it's a huge change, and cannot possibly get done in the time frame required.   you can change the block size without requiring the transactions limit (1mb max) to change... if you do that, you are more likely to succeed with an accelerated deployment.

**Comment by opetruzel** [comment]
@earonesty 
I find your argument much more compelling than anything based on BU/EC, so thank you for pointing out the issue with SPV. I hadn't considered them in this "war" of full node clients, so thank you for that reminder! You're absolutely right that we must tread more carefully, in terms of changes that SegWit2x may ultimately force on the entire ecosystem.  That said, SegWit2x may require such changes in the interest of safety and security, so there may be little choice with some changes. 

Concerning this PR, specifically, do you think the 1MB transaction cap, by itself, is sufficient to eliminate this potential attack vector?

**Comment by RHavar** [comment]
NACK. I don't think consensus rules should be added so lightly. IMO really should be something resembling a BIP to document this and if it's desired phased in as a soft fork.

**Comment by chjj** [comment]
@RHavar, the hardfork is _already_ adding a consensus rule for tx size. This simply changes that rule. Instead of a 1mb limit, it's a 1mb limit for one transaction.

**Comment by jgarzik** [comment]
@chjj The hard fork enshrines a rule that is already coded into a great many wallets.  The tx-size BIP does not change any requirements from that which is already in the field.  It just plainly states what has already occurred and already been written into the code.

**Comment by RHavar** [comment]
@chjj I don't really think there's anything wrong with this change per se, it's just we shouldn't be making consensus changes so lightly and shoe horning them into segwit2x. There's a lot of side-effects to be considered (for instance: transaction selection) that will inevitably become a pain in the ass in the future.  (For instance, is there anything now that's going to stop a miner from creating a block that accidentally breaks this rule?)

I think a BIP is the proper way to go for a consensus level change like this

**Comment by chjj** [comment]
@RHavar, this isn't me just trying to shoehorn in an unrelated change. This safety mechanism is necessary if we're going to have 2mb blocks. Given the controversy surrounding this whole situation, I highly suspect there are some well funded adversaries out there willing to burn some cash in order to try to DoS the HF network. Even 1-2% hashpower would be enough when combined with the utxo attack I described [here][1] (also note that the utxo attack does not require the participation of a miner -- sipa's pertxout branch may be required to prevent this attack).

(edit: Did github comments break?)

[1]: https://github.com/btc1/bitcoin/issues/32#issuecomment-310258839

File: src/validation.cpp [hunk 1]
```diff
@@ -3088,6 +3088,22 @@ bool ContextualCheckBlock(const CBlock& block, CValidationState& state, const Co
...
+    // Only one transaction's base size is allowed to exceed 100kb once
+    // the hardfork is activated. This exists as an extra failsafe for
+    // the O(n^2) sighashing problem. Note that the coinbase transaction
+    // is not affected by this limit.
+    if (BIP102active(nHeight, fSegwitSeasoned)) {
+        bool fSeenMaxSize = false;
+        for (size_t i = 1; i < block.vtx.size(); i++) {
+            const CTransaction &tx = *(block.vtx[i]);
+            if (::GetSerializeSize(tx, SER_NETWORK, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS) > 100000) {
+                if (fSeenMaxSize)
```
Comment (mbelshe) [inline src/validation.cpp:3100]: nit:  would prefer to see {} around the if block.  realize core doesn't do that for one-liners... :-(

**Review by mbelshe — COMMENTED** [review]
overall, agree with this PR.

**Comment by RHavar** [comment]
What are the protections in place to stop miners from creating a block that accidentals breaks this rule?

**Comment by chjj** [comment]
@RHavar, policy rules.

**Comment by jgarzik** [comment]
concept ACK - we could remove this code after the transition period

**Comment by tomasvdw** [comment]
I think this is a bad idea.

For the sake of keeping things simple, we should minimize the amount of limits. It is much easier to add limits than to remove them. The (minor) advantages of limiting all transactions to 100kb in terms of possible code optimizations and quadratic hashing do not apply with this policy.

As for the benefit. DoS attacks exploit either an exponential resource growth or an uncapped linear resource growth. Creating two 1mb transactions instead of one is _not_ a viable DoS attack.

**Comment by RHavar** [comment]
> concept ACK - we could remove this code after the transition period

With a HF? :/

I really think it's a bad idea, but if you're planning on phasing it out -- it might make sense to make the rule automatically be removed if say 2 blocks are built on top of a violating block. I think that's a can of worms, but at least it'll provide an easy upgrade path.  (And if that ever happens, you can just remove the code and pretend it never existed)

**Comment by christophebiocca** [comment]
Can we start by measuring the worst case scenario? I'm a little bit wary of declaring a DoS risk and setting up countermeasures without having a testcase on hand.

The UTXO-DoS, specifically, is entirely theoretical, and it doesn't depend on big transactions to the best of my understanding (it would be just as effective as part of a block with 1000 2KB transactions as with a block with 2 1000KB transactions).

**Comment by rfree-d** [comment]
> With a HF? :/

yo-dawg-hardfork-bitcoin

**Comment by jgarzik** [comment]
@christophebiocca More a long-validation-time / high-validation-cost type issue.

**Comment by nomnombtc** [comment]
Not sure this is a good idea to introduce more limits ... 

Maybe something like what BU does makes more sense for this problem, they have parallel processing of blocks, if one takes a long time to validate another block might orphan it in the meantime.

Edit: https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/254

**Comment by tomlevesque123** [comment]
@jgarzik "we could remove this code after the transition period" sounds like Satoshi in 2009 re: 1MB block limit. ;-)    Famous last words.

**Review by lra — CHANGES_REQUESTED** [review]
NACK

**Comment by lra** [comment]
It's not part of the Segwit2x plan. Segwit2x should not be piggybacked with every other changes that people want to add.

Such an isolated PR should be done against https://github.com/bitcoin/bitcoin and, once merged, can be merged back in btc1.

**Comment by kanoi** [comment]
If pools/miners are running some other client, that doesn't follow the rules of this client, then they may well simply be on some other fork.
Isn't the point of this project to create a client that supports /NYA/ and has HF rules a few months after it activates SegWit2x?
I'd be surprised if the aim of this project is to support non-conforming clients or other forks of Bitcoin?

Someone creating a block with 2 transactions above 100K each is questionable at best.

Any client creating those large transactions, which is not necessarily related to building blocks containing them, simply has to wait for multiple blocks to confirm them.

Personally I'd like to see a limit on the size of the single largest transaction allowed in a block.
As far as I know, the only time a 1Mb transaction has appeared was with F2Pool screwing around.
It would be way better to stop this disruptive behaviour sooner, rather than later the next time F2Pool decides to create a 2MB or 8MB transaction for the same juvenile reason.

Transaction size limits don't cripple Bitcoin, since you can simply plan to divide the transaction up into smaller transactions.

**Comment by buuser** [comment]
NACK

Transactions over 100KB are not currently relayed by the network (or >400000 tx weight) due to the code [here](https://github.com/btc1/bitcoin/blob/segwit2x/src/policy/policy.cpp#L66) . This would be a rather expensive attack to try and pull off by a miner. It could have been tried already with 1MB blocks, but we haven't seen it. 

It needs addressing in the long term, but I don't think the risk is high enough to justify having to fix it in this release.

**Comment by jgarzik** [comment]
Closing w/ rationale outlined in #32

**Pull request closed without merging**